src/util/std_expr.h

```diff
@@ -1308,6 +1308,23 @@ class equal_exprt:public binary_relation_exprt
   equal_exprt(exprt _lhs, exprt _rhs)
     : binary_relation_exprt(std::move(_lhs), ID_equal, std::move(_rhs))
   {
+    PRECONDITION(lhs().type() == rhs().type());
+  }
+
+  // Temporary workaround for callers that do not guarantee
+  // lhs.type() == rhs.type()
+  // This constructor will be removed.
+  equal_exprt(exprt _lhs, exprt _rhs, int)
+    : binary_relation_exprt(std::move(_lhs), ID_equal, std::move(_rhs))
+  {
+  }
+
+  // Temporary workaround for callers that do not guarantee
+  // lhs.type() == rhs.type()
+  // This function will be removed.
+  static equal_exprt unchecked(exprt _lhs, exprt _rhs)
+  {
+    return equal_exprt(std::move(_lhs), std::move(_rhs), 0);
   }
```

Comment:
I'd strongly suggest we fix this right away rather than introducing an API that then needs to be removed via deprecation. If there is absolutely no way of debugging this now (I'm not sure why), then we need some mechanism of tracking this.

Reply:
Tried, and gave up.

Reply:
Tracking is easy: grep for `unchecked` will do it.

Reply:
Fix the fixes that I added this isn't necessary anymore. I'll leave it to you to decide whether you want to keep these (now unused) methods.

Reply:
ok, dropped